### PR TITLE
refactor: decouple thread support from chatservice, route integrations through it

### LINKED ISSUE #612: Decouple thread support from ChatService and route messaging integrations through it
## Problem

`ChatService` (`ak-py/src/agentkernel/core/chat_service.py`) currently mixes three concerns:

1. Chat orchestration (validate, build requests, run the agent)
2. Conversation-thread recording (`_validate_thread` / `_thread_pre_run` / `_thread_post_run`)
3. HTTP-shaped presentation (`ResponseBuilder` dicts, `HTTPException`, SSE frames)

Because of this coupling:

- All seven messaging integrations (Slack, WhatsApp, Messenger, Instagram, Telegram, Teams, Gmail) bypass `ChatService` and call `AgentService` directly, so they hand-roll validation/request plumbing and silently miss every ChatService-layer feature.
- Every `ChatService` consumer unconditionally imports the full `core.thread` package, and thread support forces `user_id` on every caller when a `thread` config block exists.
- Thread support's REST routes live in the `api` package (with a config-based auto-mount in `RESTAPI.run`) instead of being packaged with the capability, unlike the messaging-integration pattern where a handler owns its routes and logic together.
- Integrations cannot use `ChatService` cleanly because its return values are HTTP-shaped payloads, not typed `AgentReply` objects.

## Product framing

Conversation threads are the history mechanism for clients that connect to the agent directly (first-party chat UIs), an **alternative** to connecting via Slack/WhatsApp/Gmail/etc., which keep their own native history. Thread support is therefore packaged, mounted, and enabled exactly like a messaging integration, and messaging integrations never record threads.

## Proposed change (three iterations, all landing in one release)

**Iteration 1: Decouple threads from ChatService, split core from presentation**
- Remove the thread linkage from `ChatService` permanently (the `core/thread/` module itself stays).
- Extract a transport-neutral execution surface (`execute` / `execute_stream`) that returns typed replies and raw `StreamChunk`s, accepts prebuilt `AgentRequest` lists, and lets exceptions propagate.
- Reimplement the existing `process_*` methods as thin presentation wrappers; REST/Lambda/ECS/Azure behavior unchanged (except thread recording, which moves in iteration 3).

**Iteration 2: Route messaging integrations through ChatService**
- All seven integration handlers stop constructing `AgentService` and call `ChatService.execute` with their prebuilt request lists.
- Platform-specific logic (attachment downloads, session-id derivation, acknowledgement flows, reply delivery, error UX) stays in the handlers; `user_id`/`group_id` populated best-effort with no consumer.
- Add handler-level unit tests (the integrations currently have none).

**Iteration 3: Rebuild thread support as an integration**
- New `integration/thread/` package owning everything thread-facing: a `ThreadRecorder` class (recording logic: pre/post run, streaming accumulation, `user_id` enforcement) and an `AgentThreadRequestHandler` that extends `AgentRESTRequestHandler` to serve the same chat routes with recording wrapped around the ChatService core, plus the thread read routes (moved out of `api/thread.py`, which is removed; breaking import-path change noted in release notes).
- Enablement follows the integration pattern: import the handler and mount it directly via `RESTAPI.run(handlers=[AgentThreadRequestHandler(...)])`. The config-based auto-mount in `RESTAPI.run` is removed; the `thread` config block only parameterizes backends/naming and no longer enables anything.
- Explicit scoping: thread recording applies only where the thread handler is mounted (the self-hosted REST API). Deployment adapters (Lambda, ECS queue and direct-WS modes, Azure) lose recording; thread-aware variants reusing `ThreadRecorder` are possible follow-ups if ever needed.

## Non-goals

- Migrating A2A, MCP, or the CLI off `AgentService` (the CLI stays on it permanently).
- Reply-type changes in integrations (replies remain stringified for delivery).
- Thread recording for messaging integrations (permanently, by product framing) and deployment adapters.
- Hook-based thread recording in `Runtime` (possible follow-up, not foreclosed by this design).
- Changes to thread stores, naming, `ConversationThreadManager`, or thread data layouts.

Design spec: `docs/specs/612-chatservice-refactor/design.md` (spec.md and plan.md follow after design review).

### PR DESCRIPTION
**Pull request by amithad**
## Description

Decouples conversation-thread support from `ChatService` and routes every messaging integration through it, per the spec set in `docs/specs/612-chatservice-refactor/` (design.md → spec.md → plan.md, reviewed iteratively).

Three-part change:

1. **ChatService execution core**: `ChatService` gains transport-neutral `execute` / `execute_sync` / `execute_stream` / `execute_stream_sync` methods returning typed replies and raw `StreamChunk`s, with support for caller-built request lists. The HTTP-shaped `process_*` methods become thin wrappers over the core with identical wire behavior. All thread linkage is removed from `ChatService` permanently.
2. **Messaging integrations on the core**: all seven handlers (Slack, WhatsApp, Messenger, Instagram, Telegram, Teams, Gmail) stop constructing `AgentService` and call `ChatService.execute` with their prebuilt request lists. Platform behavior is unchanged; dead `result.raw` branches removed; handler-level unit tests added where none existed.
3. **Threads as an integration**: the whole thread capability now lives in `integration/thread/` (public alias `agentkernel.thread`), enabled by mounting `AgentThreadRequestHandler` (chat routes with recording + read routes). The config-based auto-mount, `api/thread.py`, and `agentkernel.core.thread` are removed; the `thread` config block only parameterizes the store backend and naming.

Architecture docs and skills were synced: a chat execution layering diagram, a "which layer does new code call" rubric, a ChatService vs AgentService comparison, and a terminology sweep removing every "AgentService / ChatService" conflation.

## Type of Change

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Refactoring (no functional changes)
- [x] Documentation update
- [x] Test update
- [x] Bug fix (non-breaking change which fixes an issue)

## Related Issues

Fixes #612

## Changes Made

- `core/chat_service.py`: execution core extracted; thread methods and `ConversationThreadManager` import deleted; wrappers rebuilt over the core
- `integration/{slack,whatsapp,messenger,instagram,telegram,teams,gmail}`: `AgentService` replaced with `ChatService.execute` + `BaseChatRequest` (best-effort `user_id`/`group_id` identity mapping); `ValueError` maps to each platform's no-agent message
- `integration/thread/` (new): `AgentThreadRequestHandler`, `ThreadRecorder`, moved `ThreadRESTRequestHandler`, and the relocated thread domain (manager, models, naming, stores) from `core/thread/`; top-level `agentkernel/thread.py` alias
- `api/`: `api/thread.py` deleted, `ThreadRESTRequestHandler` export removed, config-based auto-mount removed from `RESTAPI.run`
- `core/config.py`: thread config description/docstring reflect mounting-is-enablement (no field changes; config audit found none removable)
- Examples: both thread examples mount `AgentThreadRequestHandler` from `agentkernel.thread`
- Docs site + dev skills + bundled skills: layering diagram, call rubric, ChatService vs AgentService section, threads/REST/queue-mode pages rewritten for the new model, test-table updates
- Docs site fix (independent rider commit): `toJSON` guard on `Element.prototype` in the Mermaid theme wrapper, fixing the "Converting circular structure to JSON" crash on subgraph diagrams after GSAP-animated navigation (layout-elk 0.2.2 still JSON-clones cluster nodes)
- Spec set (`docs/specs/612-chatservice-refactor/`): design.md, spec.md, plan.md, release-notes.md

## Breaking Changes

See `docs/specs/612-chatservice-refactor/release-notes.md` for paste-ready release notes.

1. Thread recording happens only via `AgentThreadRequestHandler`; deployment adapters (Lambda, ECS queue/direct-WS, Azure) no longer record threads
2. `user_id` is no longer required on non-thread chat requests (previously enforced everywhere once a `thread:` block existed)
3. Thread read routes are no longer auto-mounted from config
4. Import paths: `agentkernel.api.ThreadRESTRequestHandler` and `agentkernel.core.thread` removed; everything is importable from `agentkernel.thread`

## Testing

- [x] Unit tests pass locally (1174 passed; the only 3 failures are pre-existing `test_cli_tester.py` LLM-judge cases requiring live API access, verified failing on the pristine tree)
- [ ] Integration tests pass locally (nightly/weekly e2e workflows; not run locally)
- [x] Manual testing completed (thread example flow, docs pages with the new diagrams)
- [x] New tests added for changes (`test_chat_service_core.py` 23, `test_slack_integration.py` 10, `test_whatsapp_integration.py` 6, `test_gmail_integration.py` 6, `test_thread_integration.py` 13; `test_thread_router.py` re-pointed; `test_thread_chat_service.py` superseded and removed)

## Checklist

- [x] My code follows the project's style guidelines
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published

## Additional Notes

- All iterations land in this single PR (single release) so the intermediate state (recording removed from `ChatService` but the thread handler not yet available) never ships.
- Thread store data layouts are unchanged; threads written before this change read back identically.
- `make lint-check-all` passes (includes the Makefile lint-target fix).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: docs/specs/612-chatservice-refactor/plan.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+- `architecture/overview.md`: split the "AgentService / ChatService" node in the component diagram
+  (`:26`) and the layer table row (`:113`) into the presentation/core/AgentService layering.
+- Terminology sweep (spec: "Terminology sweep" under the layering-diagram section): rewrite the
+  conflated mentions at `execution-flow.md:11`, `core-concepts/overview.md:180`, and
+  `api/a2a-server.md:76` using the spec's canonical layer names; verify `execution-flow.md:61` and
+  the solo-ChatService pages the spec lists, updating only where behavior moved.
+- `docs/docs/integrations/*.md`: expected to need **no** update (platform setup and config are
+  unchanged); verify per page during the sync.
+
+Also: the root `AGENTS.md`/`README.md` and deployment READMEs are expected to need no update
```
Comment (amithad) [inline docs/specs/612-chatservice-refactor/plan.md:153]: **[suggestion]** `ak-py/README.md` still documents the removed thread semantics, and this surface list omits it

- `ak-py/README.md:400-412` (Conversation Thread Support) still says adding a `thread` block "turns on" threads, that `user_id` "becomes required on every chat request", and that history "becomes readable over REST" (implying the auto-mounted read routes). All three describe pre-#612 behavior.
- `ak-py/README.md:1220`: the config example's YAML comment repeats the "enables Conversation Thread Support (user_id becomes required on chat requests)" claim.
- Why: `ak-dev-sync-docs-from-branch` lists `ak-py/README.md` as a required surface, and it is the package README shown on PyPI; after this release it would contradict the shipped behavior and the rewritten `docs/docs/advanced/threads.md`.
- Suggestion: rewrite the section to the mounting model (the updated `ak-add-capabilities` SKILL.md steps are a good template) and fix the YAML comment; also add `ak-py/README.md` to this plan line so the archived plan reflects the real surface set.

File: docs/specs/612-chatservice-refactor/plan.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+## Iteration 5: Thread integration package (PR 5, `refactor:`)
+
+- **Goal:** thread support is a mountable integration; no thread code remains under `api/`.
+- **Files:** add `ak-py/src/agentkernel/integration/thread/{__init__.py,recorder.py,thread_chat.py}`
+  and `ak-py/src/agentkernel/thread.py`; relocate the whole `core/thread/` module (authoriser,
+  manager, model, naming, store/) into `integration/thread/` with import-path fixes only; delete
+  `ak-py/src/agentkernel/api/thread.py`; edit `ak-py/src/agentkernel/api/__init__.py` (drop the
+  export) and `api/http.py` (drop the auto-mount block); edit `core/config.py` (thread field
+  description and class docstring only); update
+  `examples/api/thread-openai/{app.py,app_test.py,README.md,config.yaml if needed}` and
```
Comment (amithad) [inline docs/specs/612-chatservice-refactor/plan.md:67]: **[suggestion]** The example READMEs named on this line were not updated and still describe config-presence enablement

- `examples/api/thread-openai/README.md:4-5`: "Adding a `thread` block to `config.yaml` turns on persistent conversation threads: every chat request must then carry a `user_id`". This contradicts the updated `app.py` in the same directory, whose comment says mounting `AgentThreadRequestHandler` is what enables the feature.
- `examples/api/multimodal/thread-openai/README.md:6-7` has the same framing (thread block + `multimodal.enabled` implies every chat request must carry a `user_id`).
- Suggestion: reword both intros to the mounting model. The `user_id` requirement itself stays true for these apps (all chat routes are the thread handler's), so only the enablement framing needs fixing.

File: docs/specs/612-chatservice-refactor/plan.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
+# #612 Implementation Plan: Decouple thread support from ChatService and route messaging integrations through it
+
+Five code PRs land on `develop` in strict order, all within one release (design decision: the
```
Comment (amithad) [inline docs/specs/612-chatservice-refactor/plan.md:3]: **[nit]** The rollout story is stale: this landed as one PR, not five

- The PR description states all iterations land in this single PR; `spec.md:118` similarly still says "Rollout is three PRs".
- The spec set is the durable record of the change; a future reader would conclude five PRs landed in sequence on `develop`.
- Suggestion: one line noting the actual delivery (single PR #613, single release), keeping the iteration ordering as the logical structure it remains.

**Review by amithad — COMMENTED** [review]
Review against the Agent Kernel dev skills (`ak-dev-architecture`, `ak-dev-code-quality`, `ak-dev-testing-conventions`, `ak-dev-write-spec`, `ak-dev-new-messaging-integration`, and the docs/skills sync flows). The spec set was reviewed first, then the implementation was checked against the requirements extracted from it.

**Overall: a clean, faithful implementation of a well-evidenced spec set.** Findings: 2 suggestions + 2 nits, all inline, all documentation-surface issues. No code findings survived verification.

**Spec verdict**

- Every spot-checked factual claim in `design.md`/`spec.md` verified against `develop`: the thread-method line refs, the sync/async `session_id` validation asymmetry, the sync 500-branch omitting `session_id`, both stream paths recording before agent selection (the phantom-thread claim), the dead `result.raw` branches (no `AgentReply` type has `raw`), the "no integration handler tests" claim, and all enumerated docs conflation sites.
- Each document reads at its stage's altitude; `plan.md` orders without restating and carries the tests and docs/skills-sync iterations. The 11 enumerated behavioural changes all match the implementation, and `release-notes.md` records the right four breaking changes.
- Staging is sound: `design.md` carries a "Decisions from design review" section and the PR states the set was reviewed iteratively.

**Implementation vs spec (dimension 6)**

- Requirements checklist fully implemented: core methods and validation rules, wrapper wire shapes, all seven handlers on `execute` with the spec's identity-mapping table, the thread package layout, recorder ordering and exact message strings, removals, config-description-only change, and examples' `app.py`.
- Verified fixed points: no thread residue under `core/` or `api/`; no `AgentService` reference remains in any integration; the moved `ThreadRESTRequestHandler` body is verbatim; store/manager/naming moves are import-path-only; terminology sweep is complete (no "AgentService / ChatService" conflation remains).
- New test suites match the spec's Testing section case-for-case, including the end-to-end thread read-back test and the no-phantom-thread precheck assertions.
- Deviations found are the two README surfaces flagged inline (files outside the diff, so anchored to the plan lines that name or omit them): `ak-py/README.md` and both thread example READMEs still describe config-presence enablement and the global `user_id` requirement.

**Skipped as duplicates:** Copilot already raised typing the new core methods' `requests` params as `List[AgentRequest]` (plus the missing alias import); not repeated here.

The Mermaid `toJSON` rider is declared in the PR description and is well-guarded and documented; no concern. CI: all checks green.

File: docs/specs/612-chatservice-refactor/spec.md [hunk 4]
```diff
@@ -0,0 +1,527 @@
...
+`whatsapp_chat.py:41-43`).
+
+Chat-route behavior (both `run` and `run_multipart`):
+
+1. **Agent-availability precheck** before any thread write: replicate `AgentHandler.initialize`'s rule
+   (named agent must be registered, otherwise at least one agent must exist,
+   `core/service.py:49-78`) and raise the same `ValueError("No agent available")` mapped to 400. This
+   unifies the two paths that today disagree (see Behavioural changes, item 6).
+2. Non-stream mode: `requests = await RequestBuilder.from_base_request_async(body)` (this is where
+   prompt-required and extra-field mapping run for the thread channel) →
```
Comment (amithad) [inline docs/specs/612-chatservice-refactor/spec.md]: **[nit]** `RequestBuilder.from_base_request_async` does not enforce prompt-required; the handler's own validation does

- `RequestBuilder` performs no prompt validation (it just builds `AgentRequestText(prompt=req.prompt)`); in the implementation the session_id/prompt checks live in `AgentThreadRequestHandler._validate_chat_request` (`integration/thread/thread_chat.py:169-179`), which runs before any thread write.
- The implemented behavior is correct; reword this parenthetical so the spec matches it (extra-field mapping runs in `RequestBuilder`, prompt-required runs in the handler's validation).

File: ak-py/src/agentkernel/core/chat_service.py [hunk 5]
```diff
@@ -7,6 +7,7 @@
 
 from .config import AKConfig
 from .model import (
+    AgentReply,
     AgentReplyAny,
     AgentReplyImage,
     AgentReplyText,
```
[the remark this answers is not in the document]
Comment (amithad) [reply ak-py/src/agentkernel/core/chat_service.py:13]: Done: added the AgentRequest import and typed the requests parameters as Optional[List[AgentRequest]] across execute/execute_sync/execute_stream/execute_stream_sync, plus the _prepare_async/_prepare_sync/_validate helpers (including the _prepare_* return types).

File: docs/specs/612-chatservice-refactor/plan.md (hunk 1 above)
Comment (amithad) [reply docs/specs/612-chatservice-refactor/plan.md:153]: Done: rewrote the Conversation Thread Support section of ak-py/README.md to the mounting model (AgentThreadRequestHandler serves the chat routes with recording plus the read routes; the thread block only selects backend/naming; user_id scoped to the handler's routes), fixed the YAML comment in the config example, and added ak-py/README.md to plan.md's iteration-7 surface list.

File: docs/specs/612-chatservice-refactor/plan.md (hunk 2 above)
Comment (amithad) [reply docs/specs/612-chatservice-refactor/plan.md:67]: Done: both example README intros reworded to the mounting model, pointing at the handler mount in app.py. The user_id requirement statement stays, since every chat route in these apps is the thread handler's.

File: docs/specs/612-chatservice-refactor/plan.md (hunk 3 above)
Comment (amithad) [reply docs/specs/612-chatservice-refactor/plan.md:3]: Done: added a delivery note to plan.md (shipped as single PR #613 in one release, iteration ordering preserved as the commit sequence) and reworded spec.md's 'Rollout is three PRs' line to record the actual delivery.

File: docs/specs/612-chatservice-refactor/spec.md (hunk 4 above)
Comment (amithad) [reply docs/specs/612-chatservice-refactor/spec.md]: Done: reworded the parenthetical in spec.md. Extra-field mapping runs in RequestBuilder; prompt-required is enforced by the handler's _validate_chat_request, which runs before any thread write.

**Review by malithaYL — APPROVED** [review]
